caas/kubernetes/provider/specs/decode.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package specs
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/juju/errors"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// YAMLOrJSONDecoder attempts to decode a stream of JSON documents or
+// YAML documents by sniffing for a leading { character.
+type YAMLOrJSONDecoder struct {
```

Comment:
can't we use the one from the k8s sdk?

Reply:
The reason why we have this new decoder is that they don't expose a way to enable strict mode for validation.
You can see it's still using the converter from the SDK to convert YAML to JSON then use golang JSON decoder to unmarshal the JSON string.
One thing we do here is to set `DisallowUnknownFields` before decoding.